pyogrio/tests/test_geopandas_io.py

```diff
@@ -328,11 +330,38 @@ def test_write_datetime_mixed_offset(tmp_path, use_arrow):
     assert_series_equal(result["dates"], utc_col)
 
 
+@pytest.mark.parametrize("ext", [ext for ext in ALL_EXTS if ext != ".shp"])
 @pytest.mark.filterwarnings(
     "ignore: Non-conformant content for record 1 in column dates"
 )
 @pytest.mark.requires_arrow_write_api
-def test_read_write_datetime_tz_with_nulls(tmp_path, use_arrow):
+def test_read_write_datetime_no_tz(tmp_path, ext, use_arrow):
+    dates_raw = ["2020-01-01 09:00:00.123", "2020-01-01 10:00:00"]
+    if PANDAS_GE_20:
+        dates = pd.to_datetime(dates_raw, format="ISO8601").as_unit("ms")
+    else:
+        dates = pd.to_datetime(dates_raw)
+    df = gp.GeoDataFrame(
+        {"dates": dates, "geometry": [Point(1, 1), Point(1, 1)]},
+        crs="EPSG:4326",
+    )
+    fpath = tmp_path / f"test{ext}"
+    write_dataframe(df, fpath, use_arrow=use_arrow)
+    result = read_dataframe(fpath, use_arrow=use_arrow)
+    if use_arrow and ext == ".gpkg":
+        # for GPKG with Arrow, the datetime is written as naive datetime with the
+        # correct times, but when read the naive time is assumed to be UTC, which
+        # changes the effective time so this seems wrong.
+        df["dates"] = df["dates"].dt.tz_localize(time.timezone).dt.tz_convert("UTC")
```

Comment:
```suggestion
        df["dates"] = df["dates"].dt.tz_localize("UTC")
```

I would expect this to be just interpreted as UTC (not depend on the locale timezone) ?

Reply:
This was "temporary" testing code to show the old (not ideal) behaviour. This PR now includes changes to improve the handling of dates (if GDAL >= 3.11), so code has been removed now.